File: src/src_core/fmg/core/mosaic/MosaicDrawModel.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using Fmg.Common;
using Fmg.Common.Geom;
using Fmg.Common.Notifier;
using Fmg.Core.Img;
using Fmg.Core.Types.Draw;

namespace Fmg.Core.Mosaic {

    public static class MosaicDrawModelConst {

        /// <summary> Цвет заливки ячейки по-умолчанию. Зависит от текущего UI манагера. Переопределяется одним из MVC-наследником </summary>
        public static Color DefaultBkColor = Color.Gray.Brighter();

    }

    /// <summary> MVC: draw model of mosaic field. </summary>
    /// <typeparam name="TImageInner">platform specific view/image/picture or other display context/canvas/window/panel</typeparam>
    public class MosaicDrawModel<TImageInner> : MosaicGameModel, IMosaicDrawModel<TImageInner>
        where TImageInner : class
    {

        /// <summary> Fit the area/padding to the size.
        /// <br/>
        /// При autoFit = true:
        /// <list type="number">
        /// <item><description>При любом изменении Size, Padding меняется пропорционально Size</description></item>
        /// <item><description>При любом изменении Size / MosaicType / SizeField / Padding,
        ///                    Мозаика равномерно вписывается во внутреннюю область {@link #getInnerSize()}</description></item>
        /// <item><description>При изменении Area, нужно что бы {@link #getMosaicSize()} + {@link #getPadding()}
        ///                    будут определять новый {@link #getSize()}</description></item>
        /// </list>
        ///
        /// <br/>
        /// При autoFit = false:
        /// <list type="number">
        /// <item><description>При изменении Size / MosaicType / SizeField:
        ///     <list type="bullet">
        ///     <item><description>Мозаика равномерно вписывается во вcю область {@link #getSize()} </description></item>
        ///     <item><description>при этом Padding заного перерасчитывается с нуля </description></item>
        ///     </list>
        /// </description></item>
        /// <item> <description>При изменении Offset меняется Padding так, чтобы InnerSize остался прежним </description></item>
        /// <item> <description>При изменении Padding перерасчитывается Area, так что бы мозаика вписывалась внутрь нового InnerSize. </description></item>
        /// <item> <description>При изменении Area: при этом Size и Offset не меняются, но при этом меняется Padding.left и Padding.bottom. </description></item>
        /// </list>
        /// </summary>
        private bool           _autoFit = true;
        private SizeDouble     _size;
        private BoundDouble    _padding = new BoundDouble(0);
        private TImageInner    _imgMine, _imgFlag;
        private ColorText      _colorText;
        private PenBorder      _penBorder;
        private FontInfo       _fontInfo;
        private BackgroundFill _backgroundFill;
        private Color          _backgroundColor = MosaicDrawModelConst.DefaultBkColor;
        private TImageInner    _imgBckgrnd;
        private bool lockChanging = false;

        public bool AutoFit {
            get => _autoFit;
            set { _notifier.SetProperty(ref this._autoFit, value); }
        }

        /// <summary> get mosaic size in pixels </summary>
        public SizeDouble MosaicSize => CellAttr.GetSize(SizeField);

        /// <summary> get inner size in pixels, куда равномерно вписана мозаика. Inner, т.к. снаружи есть ещё padding </summary>
        private SizeDouble InnerSize {
            get {
                var pad = Padding;
                var s = Size;
                return new SizeDouble(s.Width - pad.LeftAndRight, s.Height - pad.TopAndBottom);
            }
        }

        /// <summary> common size in pixels </summary>
        public SizeDouble Size {
            get {
                if ((_size.Width <= 0) || (_size.Height <= 0)) {
                    var s = MosaicSize;
                    var p = Padding;
                    s.Width  += p.LeftAndRight;
                    s.Height += p.TopAndBottom;
                    Size = s;
                }
                return _size;
            }
            set {
                this.CheckSize(value);
                _notifier.SetProperty(ref this._size, value);
            }
        }

        public BoundDouble Padding {
            get => _padding;
            set {
                this.CheckPadding(value);
                _notifier.SetProperty(ref this._padding, value);
            }
        }

        /// <summary> Offset to mosaic.
        /// Определяется Padding'ом  и, дополнительно, смещением к мозаике (т.к. мозаика равномерно вписана в InnerSize) </summary>
        public SizeDouble MosaicOffset {
            get {
                var pad = Padding;
                var offset     = new SizeDouble(pad.Left, pad.Top);
                var mosaicSize = MosaicSize;
                var innerSize  = InnerSize;
                if (mosaicSize == innerSize)
                    return offset;
                var dx = innerSize.Width  - mosaicSize.Width;
                var dy = innerSize.Height - mosaicSize.Height;
                return new SizeDouble(offset.Width + dx / 2, offset.Height + dy / 2);
            }
            set {
                var pad = Padding;
                var oldOffset = new SizeDouble(pad.Left, pad.Top);
                var dx = value.Width  - oldOffset.Width;
                var dy = value.Height - oldOffset.Height;
                var padNew = new BoundDouble(pad);
                padNew.Left   += dx;
                padNew.Top    += dy;
                padNew.Right  -= dx;
                padNew.Bottom -= dy;

                bool locked = lockChanging;
                try {
                    lockChanging = true;
                    Padding = padNew;
                } finally {
                    if (!locked)
                        lockChanging = false;
                }
            }
        }


        public TImageInner ImgMine {
            get => _imgMine;
            set => _notifier.SetProperty(ref _imgMine, value);
        }

        public TImageInner ImgFlag {
            get => _imgFlag;
            set =>_notifier.SetProperty(ref _imgFlag, value);
        }

        public ColorText ColorText {
            get {
                if (_colorText == null)
                    ColorText = new ColorText();
                return _colorText;
            }
            set {
                ColorText old = this._colorText;
                if (_notifier.SetProperty(ref _colorText, value)) {
                    if (old != null)
                        old.PropertyChanged -= OnColorTextPropertyChanged;
                    if (value != null)
                        value.PropertyChanged += OnColorTextPropertyChanged;
                }
            }
        }

        public PenBorder PenBorder {
            get {
                if (_penBorder == null)
                    PenBorder = new PenBorder();
                return _penBorder;
            }
            set {
                PenBorder old = this._penBorder;
                if (_notifier.SetProperty(ref _penBorder, value)) {
                    if (old != null)
                        old.PropertyChanged -= OnPenBorderPropertyChanged;
                    if (value != null)
                        value.PropertyChanged += OnPenBorderPropertyChanged;
                }
            }
        }

        public BackgroundFill BkFill {
            get {
                if (_backgroundFill == null)
                    BkFill = new BackgroundFill();
                return _backgroundFill;
            }
            set {
                var old = this._backgroundFill;
                if (_notifier.SetProperty(ref _backgroundFill, value)) {
                    if (old != null) {
                        old.PropertyChanged -= OnBackgroundFillPropertyChanged;
                        old.Dispose();
                    }
                    if (value != null)
                        value.PropertyChanged += OnBackgroundFillPropertyChanged;
                }
            }
        }

        public FontInfo FontInfo {
            get {
                if (_fontInfo == null)
                    FontInfo = new FontInfo();
                return _fontInfo;
            }
            set {
                FontInfo old = this._fontInfo;
                if (_notifier.SetProperty(ref _fontInfo, value)) {
                    if (old != null)
                        old.PropertyChanged -= OnFontInfoPropertyChanged;
                    if (value != null)
                        value.PropertyChanged += OnFontInfoPropertyChanged;
                }
            }
        }

        public Color BackgroundColor {
            get {
                return _backgroundColor;
            }
            set {
                _notifier.SetProperty(ref _backgroundColor, value);
            }
        }

        public TImageInner ImgBckgrnd {
            get => _imgBckgrnd;
            set => _notifier.SetProperty(ref _imgBckgrnd, value);
        }

        private void OnFontInfoPropertyChanged(object sender, PropertyChangedEventArgs ev) {
            _notifier.FirePropertyChanged(nameof(this.FontInfo));
        }
        private void OnBackgroundFillPropertyChanged(object sender, PropertyChangedEventArgs ev) {
            _notifier.FirePropertyChanged(nameof(this.BkFill));
        }
        private void OnColorTextPropertyChanged(object sender, PropertyChangedEventArgs ev) {
            _notifier.FirePropertyChanged(nameof(this.ColorText));
        }
        private void OnPenBorderPropertyChanged(object sender, PropertyChangedEventArgs ev) {
            _notifier.FirePropertyChanged(nameof(this.PenBorder));
        }

        protected override void OnPropertyChanged(object sender, PropertyChangedEventArgs ev) {
            base.OnPropertyChanged(sender, ev);

            if (lockChanging)
                return;

            lockChanging = true;

            // see doc for member autoFit
            try {
                if (AutoFit) {
                    // recalc padding
                    if (ev.PropertyName == nameof(this.Size)) {
                        var evEx = ev as PropertyChangedExEventArgs<SizeDouble>;
                        var oldSize = evEx.OldValue;
                        if (oldSize != default(SizeDouble))
                            Padding = this.RecalcPadding(Padding, Size, oldSize);
                    }

                    // recalc area
                    switch (ev.PropertyName) {
                    case nameof(this.Size):
                    case nameof(this.SizeField):
                    case nameof(this.MosaicType):
                    case nameof(this.Padding):
                        var innerSize = InnerSize;
                        Area = MosaicHelper.FindAreaBySize(MosaicType, SizeField, ref innerSize);
                        break;
                    }

                    // recalc size
                    if (ev.PropertyName == nameof(this.Area)) {
                        var ms = MosaicSize;
                        var p = Padding;
                        if (((ms.Width  + p.LeftAndRight) <= 0) ||
                            ((ms.Height + p.TopAndBottom) <= 0))
                        {
                            // reset padding
                            p = new BoundDouble(0);
                            Padding = p;
                        }
                        Size = new SizeDouble(ms.Width + p.LeftAndRight, ms.Height + p.TopAndBottom);
                    }
                } else {
                    // recalc area / padding
                    switch (ev.PropertyName) {
                    case nameof(this.Size):
                    case nameof(this.SizeField):
                    case nameof(this.MosaicType):
                        var s = Size;
                        var realInnerSize = s;
                        Area = MosaicHelper.FindAreaBySize(MosaicType, SizeField, ref realInnerSize);
                        var padX = (s.Width  - realInnerSize.Width ) / 2;
                        var padY = (s.Height - realInnerSize.Height) / 2;
                        Padding = new BoundDouble(padX, padY, padX, padY);
                        break;
                    }

                    // recalc area
                    if (ev.PropertyName == nameof(this.Padding)) {
                        var innerSize = InnerSize;
                        Area = MosaicHelper.FindAreaBySize(MosaicType, SizeField, ref innerSize);
                    }

                    // recalc size
                    if (ev.PropertyName == nameof(this.Area)) {
                        var sm = MosaicSize;
                        var p = Padding;
                        var s = Size;
                        Padding = new BoundDouble(p.Left, p.Top,
                                                s.Width  - sm.Width  - p.Left,
                                                s.Height - sm.Height - p.Top);
                    }
                }

                switch (ev.PropertyName) {
                case nameof(this.MosaicType):
                case nameof(this.Area):
                case nameof(this.Size):
                case nameof(this.PenBorder):
                    var penBorder = PenBorder;
                    FontInfo.Size = CellAttr.GetSq(penBorder.Width);
                    break;
                }
            } finally {
                lockChanging = false;
            }
        }

        protected override void Disposing() {
            base.Disposing();

            // unsubscribe from local notifications
            FontInfo = null;
            BkFill = null;
            ColorText = null;
            PenBorder = null;

            ImgBckgrnd= null;
            ImgFlag = null;
            ImgMine = null;
        }

    }

}

File: src/src_core/fmg/core/mosaic/MosaicDrawModel.java
package fmg.core.mosaic;

import java.beans.PropertyChangeEvent;

import fmg.common.Color;
import fmg.common.geom.BoundDouble;
import fmg.common.geom.SizeDouble;
import fmg.core.img.IImageModel;
import fmg.core.types.draw.ColorText;
import fmg.core.types.draw.FontInfo;
import fmg.core.types.draw.PenBorder;

/**
 * MVC: draw model of mosaic field.
 *
 * @param <TImageInner> platform specific view/image/picture or other display context/canvas/window/panel
 **/
public class MosaicDrawModel<TImageInner> extends MosaicGameModel implements IMosaicDrawModel<TImageInner> {

    /** Цвет заливки ячейки по-умолчанию. Зависит от текущего UI манагера. Переопределяется одним из MVC-наследником. */
    public static Color DefaultBkColor = Color.Gray().brighter();

    /** Fit the area/padding to the size.
     *
     * <ul>При autoFit = true
     * <li> При изменении Size, Padding меняется пропорционально Size
     * <li> При изменении Size / MosaicType / SizeField / Padding,
     *      Мозаика равномерно вписывается во внутреннюю область {@link #getInnerSize()}
     * <li> При изменении Area, нужно что бы {@link #getMosaicSize()} + {@link #getPadding()} будут определять новый {@link #getSize()}
     * </ul>
     *
     * <ul>При autoFit = false
     * <li> <ol>При изменении Size / MosaicType / SizeField:
     *      <li> Мозаика равномерно вписывается во вcю область {@link #getSize()}
     *      <li> при этом Padding заного перерасчитывается с нуля
     *      </ol>
     * <li> При изменении Offset меняется Padding так, чтобы InnerSize остался прежним
     * <li> При изменении Padding перерасчитывается Area, так что бы мозаика вписывалась внутрь нового InnerSize.
     * <li> При изменении Area: при этом Size и Offset не меняются, но при этом меняется Padding.left и Padding.bottom.
     * </ul>
     **/
    private boolean        _autoFit = true;
    private SizeDouble     _size;
    private BoundDouble    _padding = new BoundDouble(0);
    private TImageInner    _imgMine;
    private TImageInner    _imgFlag;
    private TImageInner    _imgBckgrnd;
    private ColorText      _colorText;
    private PenBorder      _penBorder;
    private FontInfo       _fontInfo;
    private BackgroundFill _backgroundFill;
    private Color          _backgroundColor;
    private boolean lockChanging = false;

    public static final String PROPERTY_AUTO_FIT         = "AutoFit";
    public static final String PROPERTY_IMG_MINE         = "ImgMine";
    public static final String PROPERTY_IMG_FLAG         = "ImgFlag";
    public static final String PROPERTY_IMG_BCKGRND      = "ImgBckgrnd";
    public static final String PROPERTY_COLOR_TEXT       = "ColorText";
    public static final String PROPERTY_PEN_BORDER       = "PenBorder";
    public static final String PROPERTY_BACKGROUND_FILL  = "BackgroundFill";
    public static final String PROPERTY_FONT_INFO        = "FontInfo";
    public static final String PROPERTY_BACKGROUND_COLOR = "BackgroundColor";

    @Override
    public boolean getAutoFit() { return _autoFit; }
    @Override
    public void setAutoFit(boolean autoFit) {
        _notifier.setProperty(this._autoFit, autoFit, PROPERTY_AUTO_FIT);
    }

    /** get mosaic size in pixels */
    @Override
    public SizeDouble getMosaicSize() {
        return getCellAttr().getSize(getSizeField());
    }
    /** get inner size in pixels, куда равномерно вписана мозаика. Inner, т.к. снаружи есть ещё padding */
    private SizeDouble getInnerSize() {
        BoundDouble pad = getPadding();
        SizeDouble s = getSize();
        return new SizeDouble(s.width - pad.getLeftAndRight(), s.height - pad.getTopAndBottom());
    }
    /** common size in pixels */
    @Override
    public SizeDouble getSize() {
        if ((_size == null) ||
            (_size.width <= 0) ||
            (_size.height <= 0))
        {
            SizeDouble s = getMosaicSize();
            BoundDouble p = getPadding();
            s.width  += p.getLeftAndRight();
            s.height += p.getTopAndBottom();
            setSize(s);
        }
        return _size;
    }
    @Override
    public void setSize(SizeDouble size) {
        IImageModel.checkSize(size);
        _notifier.setProperty(this._size, size, PROPERTY_SIZE);
    }

    @Override
    public BoundDouble getPadding() { return _padding; }
    @Override
    public void setPadding(BoundDouble padding) {
        IImageModel.checkPadding(this, padding);
        _notifier.setProperty(this._padding, new BoundDouble(padding), PROPERTY_PADDING);
    }

    /** Offset to mosaic.
     * Определяется Padding'ом  и, дополнительно, смещением к мозаике (т.к. мозаика равномерно вписана в InnerSize) */
    @Override
    public SizeDouble getMosaicOffset() {
        BoundDouble pad = getPadding();
        SizeDouble offset     = new SizeDouble(pad.left, pad.top);
        SizeDouble mosaicSize = getMosaicSize();
        SizeDouble innerSize  = getInnerSize();
        if (mosaicSize.equals(innerSize))
            return offset;
        double dx = innerSize.width  - mosaicSize.width;
        double dy = innerSize.height - mosaicSize.height;
        return new SizeDouble(offset.width + dx / 2, offset.height + dy / 2);
    }

    /** set offset to mosaic */
    @Override
    public void setMosaicOffset(SizeDouble offset) {
        BoundDouble pad = getPadding();
        SizeDouble oldOffset = new SizeDouble(pad.left, pad.top);
        double dx = offset.width  - oldOffset.width;
        double dy = offset.height - oldOffset.height;
        BoundDouble padNew = new BoundDouble(pad);
        padNew.left   += dx;
        padNew.top    += dy;
        padNew.right  -= dx;
        padNew.bottom -= dy;

        boolean locked = lockChanging;
        try {
            lockChanging = true;
            setPadding(padNew);
        } finally {
            if (!locked)
                lockChanging = false;
        }
    }

    @Override
    public TImageInner getImgMine() { return _imgMine; }
    public void setImgMine(TImageInner img) {
        _notifier.setProperty(this._imgMine, img, PROPERTY_IMG_MINE);
    }

    @Override
    public TImageInner getImgFlag() { return _imgFlag; }
    public void setImgFlag(TImageInner img) {
        _notifier.setProperty(this._imgFlag, img, PROPERTY_IMG_FLAG);
    }

    @Override
    public ColorText getColorText() {
        if (_colorText == null)
            setColorText(new ColorText());
        return _colorText;
    }
    @Override
    public void setColorText(ColorText colorText) {
        ColorText old = this._colorText;
        if (_notifier.setProperty(old, colorText, PROPERTY_COLOR_TEXT)) {
            if (old != null)
                old.removeListener(this::onColorTextPropertyChanged);
            if (colorText != null)
                colorText.addListener(this::onColorTextPropertyChanged);
        }
    }

    @Override
    public PenBorder getPenBorder() {
        if (_penBorder == null)
            setPenBorder(new PenBorder());
        return _penBorder;
    }
    @Override
    public void setPenBorder(PenBorder penBorder) {
        PenBorder old = this._penBorder;
        if (_notifier.setProperty(old, penBorder, PROPERTY_PEN_BORDER)) {
            if (old != null)
                old.removeListener(this::onPenBorderPropertyChanged);
            if (penBorder != null)
                penBorder.addListener(this::onPenBorderPropertyChanged);
        }
    }

    @Override
    public BackgroundFill getBackgroundFill() {
        if (_backgroundFill == null)
            setBackgroundFill(new BackgroundFill());
        return _backgroundFill;
    }
    @Override
    public void setBackgroundFill(BackgroundFill backgroundFill) {
        BackgroundFill old = this._backgroundFill;
        if (_notifier.setProperty(old, backgroundFill, PROPERTY_BACKGROUND_FILL)) {
            if (old != null)
                old.removeListener(this::onBackgroundFillPropertyChanged);
            if (backgroundFill != null)
                backgroundFill.addListener(this::onBackgroundFillPropertyChanged);
        }
    }

    @Override
    public FontInfo getFontInfo() {
        if (_fontInfo == null)
            setFontInfo(new FontInfo());
        return _fontInfo;
    }
    @Override
    public void setFontInfo(FontInfo fontInfo) {
        FontInfo old = this._fontInfo;
        if (_notifier.setProperty(old, fontInfo, PROPERTY_FONT_INFO)) {
            if (old != null)
                old.removeListener(this::onFontInfoPropertyChanged);
            if (fontInfo != null)
                fontInfo.addListener(this::onFontInfoPropertyChanged);
        }
    }

    @Override
    public Color getBackgroundColor() {
        if (_backgroundColor == null)
            setBackgroundColor(DefaultBkColor);
        return _backgroundColor;
    }

    @Override
    public void setBackgroundColor(Color color) {
        _notifier.setProperty(_backgroundColor, color, PROPERTY_BACKGROUND_COLOR);
    }

    @Override
    public TImageInner getImgBckgrnd() { return _imgBckgrnd; }
    public void setImgBckgrnd(TImageInner imgBckgrnd) {
        _notifier.setProperty(this._imgBckgrnd, imgBckgrnd, PROPERTY_IMG_BCKGRND);
    }

    private void onFontInfoPropertyChanged(PropertyChangeEvent ev) {
        _notifier.firePropertyChanged(null, ev.getSource(), PROPERTY_FONT_INFO);
    }
    private void onBackgroundFillPropertyChanged(PropertyChangeEvent ev) {
        _notifier.firePropertyChanged(null, ev.getSource(), PROPERTY_BACKGROUND_FILL);
    }
    private void onColorTextPropertyChanged(PropertyChangeEvent ev) {
        _notifier.firePropertyChanged(null, ev.getSource(), PROPERTY_COLOR_TEXT);
    }
    private void onPenBorderPropertyChanged(PropertyChangeEvent ev) {
        _notifier.firePropertyChanged(null, ev.getSource(), PROPERTY_PEN_BORDER);
    }

    @Override
    protected void onPropertyChanged(PropertyChangeEvent ev) {
        super.onPropertyChanged(ev);

        if (lockChanging)
            return;

        lockChanging = true;

        // @see javadoc {@link #autoFit}
        try {
            if (getAutoFit()) {
                // recalc padding
                if (PROPERTY_SIZE.equals(ev.getPropertyName())) {
                    SizeDouble oldSize = (SizeDouble)ev.getOldValue();
                    if (oldSize != null)
                        setPadding(IImageModel.recalcPadding(getPadding(), getSize(), oldSize));
                }

                // recalc area
                switch (ev.getPropertyName()) {
                case PROPERTY_SIZE:
                case PROPERTY_SIZE_FIELD:
                case PROPERTY_MOSAIC_TYPE:
                case PROPERTY_PADDING:
                    setArea(MosaicHelper.findAreaBySize(getMosaicType(), getSizeField(), getInnerSize(), new SizeDouble()));
                    break;
                default:
                    // none
                }

                // recalc size
                if (PROPERTY_AREA.equals(ev.getPropertyName())) {
                    SizeDouble ms = getMosaicSize();
                    BoundDouble p = getPadding();
                    if (((ms.width  + p.getLeftAndRight()) <= 0) ||
                        ((ms.height + p.getTopAndBottom()) <= 0))
                    {
                        // reset padding
                        p = new BoundDouble(0);
                        setPadding(p);
                    }
                    setSize(new SizeDouble(ms.width + p.getLeftAndRight(), ms.height + p.getTopAndBottom()));
                }
            } else {
                // recalc area / padding
                switch (ev.getPropertyName()) {
                case PROPERTY_SIZE:
                case PROPERTY_SIZE_FIELD:
                case PROPERTY_MOSAIC_TYPE:
                    SizeDouble realInnerSize = new SizeDouble();
                    SizeDouble s = getSize();
                    setArea(MosaicHelper.findAreaBySize(getMosaicType(), getSizeField(), s, realInnerSize));
                    double padX = (s.width  - realInnerSize.width ) / 2;
                    double padY = (s.height - realInnerSize.height) / 2;
                    setPadding(new BoundDouble(padX, padY, padX, padY));
                    break;
                default:
                    // none
                }

                // recalc area
                if (PROPERTY_PADDING.equals(ev.getPropertyName()))
                    setArea(MosaicHelper.findAreaBySize(getMosaicType(), getSizeField(), getInnerSize(), new SizeDouble()));

                // recalc size
                if (PROPERTY_AREA.equals(ev.getPropertyName())) {
                    SizeDouble sm = getMosaicSize();
                    BoundDouble p = getPadding();
                    SizeDouble s = getSize();
                    setPadding(new BoundDouble(p.left, p.top,
                                               s.width  - sm.width  - p.left,
                                               s.height - sm.height - p.top));
                }
            }

            // change font size (пересчитать и установить новую высоту шрифта)
            switch (ev.getPropertyName()) {
            case MosaicGameModel.PROPERTY_MOSAIC_TYPE:
            case MosaicGameModel.PROPERTY_AREA:
            case MosaicDrawModel.PROPERTY_SIZE:
            case MosaicDrawModel.PROPERTY_PEN_BORDER:
                PenBorder penBorder = getPenBorder();
                getFontInfo().setSize(getCellAttr().getSq(penBorder.getWidth()));
                break;
            default:
                // none
            }
        } finally {
            lockChanging = false;
        }
    }

    @Override
    public void close() {
        super.close();
        if (_backgroundFill != null)
            getBackgroundFill().close();
        // unsubscribe from local notifications
        setFontInfo(null);
        setBackgroundFill(null);
        setColorText(null);
        setPenBorder(null);

        setImgBckgrnd(null);
        setImgFlag(null);
        setImgMine(null);
    }

}
